optimize FloatToFused8BitRowwiseQuantized and Fused8BitRowwiseQuantizedToFloat

jspark1105 opened the pull request:
Summary:
Optimize performance of these two operators.
Additionally use nearbyint instead of round to be consistent with 4-bit embedding table quantization.

Reviewed By: hyuen

Differential Revision: D19072103

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 2b0a598:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=2b0a5982a5e984c31ca08adee8e3098a2f0cc1aa).

### :detective: 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4026983) pytorch_windows_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=77987506) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=77987506))
<details>
<summary>
<code>FAILED: caffe2/CMakeFiles/torch_cpu.dir/operators/fused_rowwise_8bit_conversion_ops.cc.obj</code>
</summary>

```

_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /Z7 /EHa /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -O2 -DCAFFE2_BUILD_MAIN_LIB -DONNX_BUILD_MAIN_LIB -std:c++14 /showIncludes /Focaffe2\CMakeFiles\torch_cpu.dir\operators\generate_proposals_op.cc.obj /Fdcaffe2\CMakeFiles\torch_cpu.dir\ /FS -c ..\caffe2\operators\generate_proposals_op.cc 
Microsoft (R) C/C++ Optimizing Compiler Version 19.23.28107 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

ental -DHAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /Z7 /EHa /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -O2 -DCAFFE2_BUILD_MAIN_LIB -DONNX_BUILD_MAIN_LIB -std:c++14 /showIncludes /Focaffe2\CMakeFiles\torch_cpu.dir\operators\group_norm_op.cc.obj /Fdcaffe2\CMakeFiles\torch_cpu.dir\ /FS -c ..\caffe2\operators\group_norm_op.cc 
Microsoft (R) C/C++ Optimizing Compiler Version 19.23.28107 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

VX2_CPU_DEFINITION /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /Z7 /EHa /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -O2 -DCAFFE2_BUILD_MAIN_LIB -DONNX_BUILD_MAIN_LIB -std:c++14 /showIncludes /Focaffe2\CMakeFiles\torch_cpu.dir\operators\fused_rowwise_8bit_conversion_ops.cc.obj /Fdcaffe2\CMakeFiles\torch_cpu.dir\ /FS -c ..\caffe2\operators\fused_rowwise_8bit_conversion_ops.cc 
FAILED: caffe2/CMakeFiles/torch_cpu.dir/operators/fused_rowwise_8bit_conversion_ops.cc.obj  
VX2_CPU_DEFINITION /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /Z7 /EHa /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -O2 -DCAFFE2_BUILD_MAIN_LIB -DONNX_BUILD_MAIN_LIB -std:c++14 /showIncludes /Focaffe2\CMakeFiles\torch_cpu.dir\operators\fused_rowwise_8bit_conversion_ops.cc.obj /Fdcaffe2\CMakeFiles\torch_cpu.dir\ /FS -c ..\caffe2\operators\fused_rowwise_8bit_conversion_ops.cc 
C:\Users\circleci\project\caffe2/operators/fused_rowwise_8bit_conversion_ops.h(110): error C2398: Element '2': conversion from 'unsigned __int64' to '_Ty' requires a narrowing conversion
        with
        [
            _Ty=int64_t
        ]
C:\Users\circleci\project\caffe2/operators/fused_rowwise_8bit_conversion_ops.h(98): note: while compiling class template member function 'bool caffe2::Fused8BitRowwiseQuantizedToFloatOp<c10::Half,void caffe2::`anonymous-namespace'::convertfp32fp16(c10::Half *,const float *,size_t),true,caffe2::CPUContext>::RunOnDevice(void)'
C:\Users\circleci\project\c10/util/Registry.h(184): note: see reference to class template instantiation 'caffe2::Fused8BitRowwiseQuantizedToFloatOp<c10::Half,void caffe2::`anonymous-namespace'::convertfp32fp16(c10::Half *,const float *,size_t),true,caffe2::CPUContext>' being compiled
..\caffe2\operators\fused_rowwise_8bit_conversion_ops.cc(129): note: see reference to function template instantiation 'ObjectPtrType c10::Registerer<std::string,ObjectPtrType,const caffe2::OperatorDef &,caffe2::Workspace *>::DefaultCreator<caffe2::Fused8BitRowwiseQuantizedToFloatOp<c10::Half,void caffe2::`anonymous-namespace'::convertfp32fp16(c10::Half *,const float *,size_t),true,caffe2::CPUContext>>(const caffe2::OperatorDef &,caffe2::Workspace *)' being compiled
        with
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

jspark1105 committed:
optimize FloatToFused8BitRowwiseQuantized and Fused8BitRowwiseQuantizedToFloat (#31470)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/31470

Optimize performance of these two operators.
Additionally use nearbyint instead of round to be consistent with 4-bit embedding table quantization.

Reviewed By: hyuen

Differential Revision: D19072103

fbshipit-source-id: 87c5b8646471a27645779097743c442d8165b6cb
